### PR TITLE
Reduce persist execution

### PR DESCRIPTION
**Pull request by silvexis**
### Description
Tune Persist to reduce the number of lambda invokes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -124,16 +124,16 @@ sls: guard-action
 	@if [ -z "$${stage}" ] ; then make guard-stage ; make sls_help ; exit 1 ; fi
 	@if [ -n "$${AWS_ACCESS_KEY_ID}" -a -n "$${AWS_SECRET_ACCESS_KEY}" ] ; then \
         printf "$(WARN_COLOR)WARN:$(NO_COLOR) Found AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY vars, using those for sls "$${action}".\n" ; \
-        sls "$${action}" --stage=$${stage} --kinesis=$${kinesis:-false} ; \
+        sls "$${action}" --stage=$${stage} --kinesis=$${kinesis:-true} ; \
```
Comment (silvexis) [inline Makefile]: unrelated change to enable kinesis by default

File: reactor/features/accounts/verify.py [hunk 2]
```diff
@@ -192,8 +192,7 @@ def _do_verification(details):
             verify_responses['cloudtrail_sqs'] = list(aws_tools.verify_sqs_access(aws,
                                                                                   details['sqs_arn'].split(':')[-1]))
             policy_role_arn = details['role_arn']
-            policy_values_to_check = [details['sqs_arn'],
-                                      details['cloudtrail_bucket']]
+            policy_values_to_check = [details['sqs_arn']]
```
Comment (silvexis) [inline reactor/features/accounts/verify.py]: fix unrelated bug

File: reactor/features/resources/persist.py [hunk 3]
```diff
@@ -13,30 +13,46 @@
...
-    resource_object_key = f'resources/{environment_id}/{res_id_hash}.gz'
-    resource_object_value = utils.encode_and_compress(json.dumps(resource_data))
-    logger.info(f"Persisting '{bucket_name}/{resource_object_key}'")
-    logger.debug(resource_data)
-    Provider(service='s3').client.put_object(Bucket=bucket_name, Key=resource_object_key,
-                                             Body=resource_object_value, ServerSideEncryption='AES256')
-    return bucket_name, resource_object_key
+    persist_count = 0
+    payload_size = len(json.dumps(payload))
+    for res_id, resource in payload.items():
```
Comment (silvexis) [inline reactor/features/resources/persist.py]: persist now takes a list of records to write vs. writing them one per invoke which was rather inefficient

File: serverless.yml [hunk 4]
```diff
@@ -30,7 +30,7 @@ provider:
   apiKeys:
     - cloudzero-reactor-${self:provider.stage}-api-key
   runtime: python3.6
-  timeout: 150
+  timeout: 75
```
Comment (silvexis) [inline serverless.yml]: reduced default timeout to something smaller to limit costs when lambdas run away. This seemed to be happening today a bunch because of cloud weather

File: reactor/features/resources/curate.py [hunk 5]
```diff
@@ -201,11 +201,7 @@ def curate(records, reactor_account_id, namespace):
 
     start = timeit.default_timer()
     logger.info(f"Persisting {len(resources_to_write)} resources out of {event_stats['coresync']} to s3")
-    for res_id, resource in resources_to_write.items():
-        payload = {'res_id': res_id,
-                   'environment_id': resource['environment_id'],
-                   'resource_data': resource['data_image']}
-        lambda_utils.reactor_invoke(namespace, 'resources-persist', payload=json.dumps(payload), async=True)
+    lambda_utils.reactor_invoke(namespace, 'resources-persist', payload=json.dumps(resources_to_write), async=True)
```
Comment (killthrush) [inline reactor/features/resources/curate.py]: seems legit.  did this improve the overall execution time?

File: reactor/features/resources/persist.py [hunk 6]
```diff
@@ -13,30 +13,46 @@
...
+        resource_object_key = f'resources/{environment_id}/{res_id_hash}.gz'
+        resource_object_value = utils.encode_and_compress(json.dumps(resource_data))
+        logger.debug(f"Persisting '{bucket_name}/{resource_object_key}'")
+        Provider(service='s3').client.put_object(Bucket=bucket_name, Key=resource_object_key,
+                                                 Body=resource_object_value, ServerSideEncryption='AES256')
+        persist_count += 1
+
+    # no idea exactly why lambdas get called sometimes with no payload
+    if persist_count > 0:
+        logger.info(f"Persisted {persist_count} records to s3 ({payload_size:,} bytes)")
```
Comment (killthrush) [inline reactor/features/resources/persist.py]: Hmmm.. looks like this isn't one of our pingable lambdas...  My guess is that errors upstream might be to blame.  It's possible given the code.  Suggest avoiding the invoke entirely if there isn't anything to pass over

File: serverless.yml [hunk 7]
```diff
@@ -610,8 +611,7 @@ resources:
   # Resources END
   # Conditions BEGIN
   Conditions:
-    # only conditionally enable the kinesis stream resource,
-    # because it is quite expensive.
+    # Conditionally enable the kinesis stream resource
```
Comment (killthrush) [inline serverless.yml]: I feel like this should be added to the readme before you merge

File: reactor/features/resources/curate.py (hunk 5 above)
Comment (silvexis) [reply reactor/features/resources/curate.py]: yeah, for an average set of records, it would have ~25 lambdas executing each at about 200 to 400 ms execution vs 1 lambda executing with an average of about 800ms

File: reactor/features/resources/persist.py (hunk 6 above)
Comment (silvexis) [reply reactor/features/resources/persist.py]: ahhh doh of course, that's where the calls are coming from...will fix
